### PR TITLE
#3305 - fix default CharSequence serialization

### LINKED ISSUE #3305: ObjectMapper serializes `CharSequence` subtypes as POJO instead of as String (JDK 15+)
**Describe the bug**
I have a class that implements CharSequence and in Java 14 and below the class was serialised correctly.

 In Java 15 the JSON generated is 

{"empty": false}

I believe the change was introduced in the CharSequence API by the introduction of the method isEmpty(). 

It appears the POJOPropertiesCollector._addGetterMethod collects the method isEmpty(); 

The class is then identified as a Bean type object, and uses a BeanSerializer to serialize the class. In Java 14 ToStringSerializer would be used to serialize the class.

**Version information**
Jackson 2.11.3

**To Reproduce**

See simple class to reproduce:

`
public class TestMapper
{
    
    private final static ObjectMapper mapper;     

    static
    {
        mapper = new ObjectMapper();
        mapper.disable(SerializationFeature.FAIL_ON_EMPTY_BEANS);
        mapper.setSerializationInclusion(JsonInclude.Include.NON_NULL);
     
    }
    
    
    private static class TestObj implements CharSequence, Serializable
    {

        private String str;
        
        private TestObj(String str)
        {
            this.str = str;
        }
        
        @Override
        public int length()
        {
            return str == null ? 0 : str.length();
        }

        @Override
        public char charAt(int index)
        {
            return str == null ? 0 : str.charAt(index);
        }

        @Override
        public CharSequence subSequence(int start, int end)
        {
            return str == null ? null : str.subSequence(start, end);
        }

        @Override
        public String toString()
        {
            return str;
        }
        
    }

    public static void main(String[] args)
    {
        TestObj obj = new TestObj("Hello World");
        StringWriter writer = new StringWriter();
        try
        {
            mapper.writeValue(writer, obj);
            System.out.println(writer.toString());
        }
        catch(Exception e)
        {
            //this should not throw
        }
    }
}

`

**Expected behaviour**
Expected "Hello World"

**Actual**
{"empty": false}

Running in Java 14 or below the expected result is returned

### PR DESCRIPTION
**Pull request by seregamorph**
Fixes #3305 and #3331 (dup).

### CODE REVIEW

File: src/main/java/com/fasterxml/jackson/databind/ser/BeanSerializerFactory.java [hunk 1]
```diff
@@ -767,6 +769,28 @@ protected void removeSetterlessGetters(SerializationConfig config, BeanDescripti
         }
     }
 
+    /**
+     * Helper method that will remove all properties that are added by default JDK methods,
+     * e.g. java.lang.CharSequence.isEmpty() (since JDK 15).
+     */
+    private void removeDefaultProperties(List<BeanPropertyWriter> properties)
```
Comment (seregamorph) [inline src/main/java/com/fasterxml/jackson/databind/ser/BeanSerializerFactory.java:776]: I'm not quite sure about the policy with new protected methods, so I minimized the visibility for the first iteration

**Comment by cowtowncoder** [comment]
Thank you @seregamorph! I think the idea of the fix is sound -- I may want to tweak details a bit, but this makes sense to me. I'll add this on my TODO list: lately my time on Jackson has been severely limited and so it may take a while to get this processed, but I wanted to add a note that it is on my radar.

**Comment by seregamorph** [comment]
Sure, thanks.

**Comment by seregamorph** [comment]
@cowtowncoder I have a feeling that you plan to release the version soon with CVE. Can you please include this fix as well? We need it in our project. Thanks

**Comment by cowtowncoder** [comment]
@seregamorph Yes, I will definitely get this fix included first. It is next one on my list.

**Comment by cowtowncoder** [comment]
Thank you @seregamorph! I used this approach, only changing some of minor details (after checking if there might be other places to get rid of the property, finding no better one) -- thank you for suggesting the fix. I'll close the PR; fix will go in 2.13.1, which should be released some time this month (December 2021).

**Comment by seregamorph** [comment]
@cowtowncoder great, thanks!
Two questions:
1) Do you plan to backport it to 2.12.x as well? There is at least one good reason for it - [spring-boot 2.5.7](https://search.maven.org/artifact/org.springframework.boot/spring-boot-dependencies/2.5.7/pom) depends on `jackson 2.12.5` and on the same time spring-boot 2.5 is compatible with JDK 17. `2.5.x` will probably be yet updated and I don't think they will do major version upgrade in scope of 2.5.x.
2) I'd recommend to include the test that I made as well (or something like it), in the future it can prevent possible regression.

**Comment by cowtowncoder** [comment]
I wasn't think of backporting, but maybe I should. Hard to gauge how important the fix is, vs. any potential regression.

As to test, yes, I added it, I think under `src/test-jdk14`. Although I realize now that it need not actually go there as it does not actually need more recent jdks (jdk14 just happened to already exist)

**Comment by cowtowncoder** [comment]
Patch was included in:

* `2.13.1`
* `2.12.6`

**Pull request closed without merging**